src/nvim/api/private/helpers.c

```diff
@@ -476,6 +476,34 @@ bool buf_collect_lines(buf_T *buf, size_t n, int64_t start, bool replace_nl, Arr
   return true;
 }
 
+/// lua specialized version of buf_collect_lines that avoids string allocations
+bool buf_collect_lines_lua(buf_T *buf, size_t n, int64_t start, lua_State *lstate, Error *err)
+{
+  lua_createtable(lstate, (int)n, 0);
+
+  for (linenr_T i = 0; i < (linenr_T)n; i++) {
+    linenr_T lnum = (linenr_T)start + i;
+
+    if (lnum >= MAXLNUM) {
+      if (err != NULL) {
+        api_set_error(err, kErrorTypeValidation, "Line index is too high");
+      }
+      return false;
+    }
+
+    char *bufstr = (char *)ml_get_buf(buf, lnum, false);
+    size_t len = strlen(bufstr);
+
+    // Vim represents NULs as NLs
+    strchrsub(bufstr, '\n', '\0');
```

Comment:
This is unsafe as `ml_get_buf` is allowed to return a direct pointer to internal memline data.